LifeSpace/Debug/LogViewer.swift

```diff
@@ -0,0 +1,112 @@
+//
+//  LogShareView.swift
+//  LifeSpace
+//
+//  Created by Vishnu Ravi on 11/8/24.
+//
+
+import OSLog
+import Spezi
+import SwiftUI
+
+
+struct LogViewer: View {
+    @Environment(LogManager.self) var manager
+    
+    @State private var startDate: Date = Calendar.current.date(byAdding: .day, value: -1, to: Date()) ?? Date()
+    @State private var endDate = Date()
+    @State private var selectedLogType: LogType = .all
+    @State private var logs = ""
+    @State private var isLoading = false
+    
+    var body: some View {
+        VStack {
+            /// Date range selection
+            HStack {
+                DatePicker("FROM", selection: $startDate, displayedComponents: .date)
+                Spacer()
+                DatePicker("TO", selection: $endDate, displayedComponents: .date)
+            }
+            .padding()
+            
+            /// Log type selection
+            Picker("LOG_TYPE", selection: $selectedLogType) {
+                ForEach(LogType.allCases) { type in
+                    Text(type.rawValue).tag(type)
+                }
+            }
+            .pickerStyle(SegmentedPickerStyle())
+            .padding()
+            
+            ScrollView {
+                if isLoading {
+                    ProgressView("LOADING_LOGS")
+                        .padding()
+                } else {
+                    Text(logs)
+                        .padding()
+                }
+            }
+        }
+        .navigationTitle("LOG_VIEWER")
+        .onAppear {
+            Task {
+                await queryLogs()
+            }
+        }
+        .onChange(of: startDate) {
+            Task {
+                await queryLogs()
+            }
+        }
+        .onChange(of: endDate) {
+            Task {
+                await queryLogs()
+            }
+        }
+        .onChange(of: selectedLogType) {
+            Task {
+                await queryLogs()
+            }
+        }
+        .toolbar {
+            if !logs.isEmpty, let fileURL = saveLogsToTextFile(logs) {
+                ShareLink(
+                    item: fileURL,
+                    preview: SharePreview(
+                        "LOGS",
+                        image: Image(systemName: "doc.text") // swiftlint:disable:this accessibility_label_for_image
+                    )
+                ) {
+                    Image(systemName: "square.and.arrow.up") // swiftlint:disable:this accessibility_label_for_image
+                }
+            }
+        }
+    }
+    
+    @MainActor
+    private func queryLogs() async {
+        isLoading = true
+        
+        /// This is very slow, so run as a detached task with high priority
+        logs = await Task.detached(priority: .userInitiated) { [manager, startDate, endDate, selectedLogType] in
+            await manager.query(startDate: startDate, endDate: endDate, logType: selectedLogType.osLogLevel)
+        }.value
```

Comment:
You might want to store the task here and cancel it in case the function is called quickly after reach other to avoid race conditions and executing multiple ones after each other.

Reply:
Thanks - I've updated this to store the task and cancel it if needed:

```
    private func queryLogs() {
        /// Cancel any existing query task
        queryTask?.cancel()
        
        /// Set loading state
        isLoading = true
        
        /// Create a new query task and store it
        queryTask = Task(priority: .userInitiated) { [manager, startDate, endDate, selectedLogType] in
            /// Run the query
            let result = await manager.query(startDate: startDate, endDate: endDate, logType: selectedLogType.osLogLevel)
            
            /// Check to make sure the task isn't cancelled before updating UI
            guard !Task.isCancelled else {
                return
            }
            
            /// Update the UI
            await MainActor.run {
                logs = result
                isLoading = false
            }
        }
    }
```